### PR TITLE
Websocket

### PR DESCRIPTION
**Pull request by nkuba**
This PR:
- implements web socket client
- adds linting
- fixes protocol methods
- adds simple tests
- refactors and cleans up code
- comments out persistency strategy due to instability on testing

TODO:
https://github.com/nkuba/electrum-client-js/issues

Closes: https://github.com/nkuba/electrum-client-js/issues/7, https://github.com/nkuba/electrum-client-js/issues/6

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -1,21 +1,31 @@
...
+    "lint:fix": "eslint --fix ."
+  },
+  "dependencies": {
+    "websocket": "^1.0.29"
   },
-  "dependencies": {},
   "devDependencies": {
-    "electrum-host-parse": "*"
+    "chai": "^4.2.0",
+    "electrum-host-parse": "*",
```
Comment (liamzebedee) [inline package.json]: Add a version here, I see `^0.1.1` is latest

File: src/socket/socket_client_ws.js [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+    this.port = port
+    this.protocol = protocol
+    this.options = options
+    this.client = null
+  }
+
+  async connect() {
+    const url = `${this.protocol}://${this.host}:${this.port}`
+
+    const client = new W3CWebSocket(
```
Comment (liamzebedee) [inline src/socket/socket_client_ws.js]: Placeholder doc link for now https://github.com/theturtle32/WebSocket-Node/blob/master/docs/W3CWebSocket.md#constructor

File: src/socket/socket_client.js [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+
+class SocketClient {
+  constructor(host, port, protocol, options) {
+    this.id = 0
+    this.host = host
+    this.port = port
+    this.protocol = protocol
+    this.options = options
+    this.callback_message_queue = {}
+    this.subscribe = new EventEmitter()
```
Comment (liamzebedee) [inline src/socket/socket_client.js]: I think that this makes more sense to be `.events`, based on its type. What's the rationale for the rename?

File: package.json (hunk 1 above)
Comment (liamzebedee) [reply package.json]: Still unaddressed

Comment (nkuba) [reply package.json]: Yes, I'm not sure if we need this dependency at all, it's used only in examples. I have no decision on this yet.

File: src/socket/socket_client.js (hunk 3 above)
Comment (nkuba) [reply src/socket/socket_client.js]: Yeah, I like `events` better too. But we had to fork a base library where it's `subscribe`. I didn't want to diverge too much from the base repo, but later refactored it significantly anyways so I think this rename from `subscribe` to `events` won't be a huge change compared to all others. 
I'd like to propose we do this rename later, as we have the current solution tested and confirmed on both your and my machines. I plan to do more refactoring after the demo so will pick up this rename too.

File: package.json (hunk 1 above)
Comment (nkuba) [reply package.json]: Version set.

File: src/socket/socket_client.js (hunk 3 above)
Comment (nkuba) [reply src/socket/socket_client.js]: Created an issue for this change: https://github.com/nkuba/electrum-client-js/issues/3

**Comment by nkuba** [comment]
@liamzebedee it's ready for another chance.

File: package.json [hunk 4]
```diff
@@ -1,21 +1,31 @@
 {
-  "name": "electrum-client",
-  "version": "0.0.7",
-  "description": "Electrum protocol client for node.js",
+  "name": "electrum-client-js",
+  "version": "0.1.0",
+  "description": "Electrum protocol client for JS",
```
Comment (liamzebedee) [inline package.json]: Technically it's for the browser :)

File: test/integration_test.js [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+      await client.close()
+    })
+
+    it('blockchain_transaction_get', async () => {
+      const expoectedResult = txData.hex
+      const result = await client.blockchain_transaction_get(txData.hash)
+
+      assert.equal(
+        result,
+        expoectedResult,
```
Comment (liamzebedee) [inline test/integration_test.js]: Minor typo here

File: test/config.js [hunk 6]
```diff
@@ -0,0 +1,26 @@
...
+
+const ws = {
+  protocol: 'ws', port: '50003', host: 'electrumx-server.tbtc.svc.cluster.local',
+}
+
+const wss = {
+  protocol: 'wss', port: '50004', host: 'electrumx-server.tbtc.svc.cluster.local', options: {agent: new https.Agent({rejectUnauthorized: false})},
+}
+
+module.exports = [
```
Comment (liamzebedee) [inline test/config.js]: If we're exporting an array/list, I think it'd be nice to name this "configs". Thoughts?

File: test/integration_test.js [hunk 7]
```diff
@@ -14,33 +14,54 @@ describe('ElectrumClient', async () => {
...
-    })
+      after(async () => {
+        await client.close()
+      })
 
-    it('blockchain_transaction_get', async () => {
-      const expoectedResult = txData.hex
-      const result = await client.blockchain_transaction_get(txData.hash)
+      it('request returns result', async () => {
+        const expoectedResult = txData.hex
```
Comment (liamzebedee) [inline test/integration_test.js]: Minor typo here

File: example/bitcoin.js [hunk 8]
```diff
@@ -1,24 +1,24 @@
-const ElectrumClient = require("..")
+const ElectrumClient = require('..')
```
Comment (liamzebedee) [inline example/bitcoin.js]: I ran this test and it seems it's broken?

```
(base) ➜  electrum-client-js git:(websocket) ✗ node example/bitcoin.js 
begin connection: {"host":"electrum.vom-stausee.de","ssl":50002,"tcp":50001,"pruning":null,"http":null,"https":null}
node[8736]: ../src/tcp_wrap.cc:279:static void node::TCPWrap::Connect(const FunctionCallbackInfo<v8::Value> &): Assertion `args[2]->IsUint32()' failed.
 1: 0x10003b540 node::Abort() [/usr/local/bin/node]
 2: 0x10003a638 node::AddEnvironmentCleanupHook(v8::Isolate*, void (*)(void*), void*) [/usr/local/bin/node]
 3: 0x1000dc347 node::TCPWrap::Connect(v8::FunctionCallbackInfo<v8::Value> const&) [/usr/local/bin/node]
 4: 0x1001df32f v8::internal::FunctionCallbackArguments::Call(v8::internal::CallHandlerInfo*) [/usr/local/bin/node]
 5: 0x1001de929 v8::internal::MaybeHandle<v8::internal::Object> v8::internal::(anonymous namespace)::HandleApiCallHelper<false>(v8::internal::Isolate*, v8::internal::Handle<v8::internal::HeapObject>, v8::internal::Handle<v8::internal::HeapObject>, v8::internal::Handle<v8::internal::FunctionTemplateInfo>, v8::internal::Handle<v8::internal::Object>, v8::internal::BuiltinArguments) [/usr/local/bin/node]
 6: 0x1001de0e5 v8::internal::Builtin_Impl_HandleApiCall(v8::internal::BuiltinArguments, v8::internal::Isolate*) [/usr/local/bin/node]
 7: 0x5ffa0f5be3d 
[1]    8736 abort      node example/bitcoin.js
```

Possible that it's a result of npm deps and this new machine I've got. But I did a clone of the base repo, and it seemed to work-

```
begin connection: {"host":"electrum.hsmiths.com","ssl":50002,"tcp":50001,"pruning":null,"http":null,"https":null}
\{ code: 1, message: 'unsupported protocol version: 1.0' }
OnClose:undefined
```

Do we intend to support the old Node.js part of this library, or just want to go full hog on implementing pure browser JS? 😄

File: package.json (hunk 4 above)
Comment (nkuba) [reply package.json]: It's for both browser and node.js.

File: example/bitcoin.js (hunk 8 above)
Comment (nkuba) [reply example/bitcoin.js]: In the new version, we support node.js and browser. We need to update examples scripts to reflect the improvements we introduced.

Comment (nkuba) [reply example/bitcoin.js]: Created a separate issue for that: https://github.com/nkuba/electrum-client-js/issues/7

File: test/config.js (hunk 6 above)
Comment (nkuba) [reply test/config.js]: I modified it a little bit, please take a look.

File: package.json (hunk 4 above)
Comment (liamzebedee) [reply package.json]: Ahhh. Can you add "browser and node.js" in the description. Some projects get lazy and JS doesn't always mean for Node.js and vice-versa :)

File: test/config.js [hunk 9]
```diff
@@ -1,26 +1,29 @@
...
+  // TODO: WebSocket is commented out for CI, until we find public servers for this protocol.
+  // electrumServers.ws,
+  // electrumServers.wss,
+]
 
-const wss = {
-  protocol: 'wss', port: '50004', host: 'electrumx-server.tbtc.svc.cluster.local',
+module.exports = {
+  electrumServers,
+  serversArray,
```
Comment (liamzebedee) [inline test/config.js]: I think we could just rename this to `servers`. Thoughts?

File: test/config.js [hunk 10]
```diff
@@ -1,26 +1,29 @@
 /* eslint no-unused-vars: 0 */
 
-const tcp = {
-  protocol: 'tcp', port: '50001', host: 'fortress.qtornado.com',
+const electrumServers = {
```
[the remark this answers is not in the document]
Comment (nkuba) [reply test/config.js]: Good idea! We will leave it for now as is, we haven't focused on test much, yet.

Looking forward to hearing more about your findings around testnet infrastructure.

**Pull request merged**